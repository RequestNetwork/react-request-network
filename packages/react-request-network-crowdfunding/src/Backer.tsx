import * as React from 'react';
import { Consumer } from "@requestnetwork/react-components";

interface IProps {
  requestId: string;
  requestNetwork: any;
  component: any;
}

export class InnerBacker extends React.Component<IProps> {
  public state = {
    backed: false,
    metaMaskLoading: false,
  };
  public handleSubmit = async (amount:number, requestId:string, requestNetwork) => {
    this.setState({ metaMaskLoading: true });
    const amountInWei = amount * Math.pow(10, 18);
    requestNetwork
      .pay(requestId, amountInWei.toString())
      .on('broadcasted', () =>
        this.setState({ backed: true, metaMaskLoading: false })
      )
      .catch(() => this.setState({ metaMaskLoading: false }));
  };

  public render() {
    const { component, requestNetwork, requestId } = this.props;
    const { metaMaskLoading, backed } = this.state;
    const props = {
      backed,
      metaMaskLoading,
      disabled: !requestId,
      isReady: requestNetwork.isReady,
      metaMaskDisabled: !requestNetwork.currentAccount,
      pay: (amount:number) => this.handleSubmit(amount, requestId, requestNetwork),
    };
    return React.createElement(component, props);
  }
}
const Backer = props => (
  <Consumer>
    {requestNetwork => (
      <InnerBacker {...props} requestNetwork={requestNetwork} />
    )}
  </Consumer>
);
export default Backer;
