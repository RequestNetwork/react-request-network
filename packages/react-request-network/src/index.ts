import * as React from "react";
import RequestNetworkProvider from "./Provider";
import { Creator } from "./Creator";

interface ContextValues {
  isReady: boolean;
  networkMismatch: boolean;
  currentNetwork: string;
  currentAccount: string;
  create(paymentAddress: string, amount: string, data: any);
  get(id);
  pay(requestId: string, amount: string);
}

export { Creator };
export const { Provider, Consumer } = React.createContext(<ContextValues>{
  isReady: false,
  networkMismatch: false,
  currentNetwork: "",
  currentAccount: ""
});

export default RequestNetworkProvider;
