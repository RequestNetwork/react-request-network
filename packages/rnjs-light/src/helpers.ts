import { default as requestArtifact } from "requestnetworkartifacts";

const defaultNetwork = "private";
const defaultAddress = "last-RequestEthereum";

export const getArtifact = (
  network = defaultNetwork,
  address = defaultAddress
) => requestArtifact.default(network, address);
