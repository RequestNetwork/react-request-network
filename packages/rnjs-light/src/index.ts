const EthereumTx = require("ethereumjs-tx");
const Contract = require("web3-eth-contract");
const { getArtifact } = require("./utils");
const ethers = require("ethers");

// example privateKey
const privateKey = Buffer.from(
  "c87509a1c067bbde78beb793e6fa76530b6382a4c0241e5e4a9ec0a0f44dc0d3",
  "hex"
);

const artifact = getArtifact();
const contractAddress = artifact.networks.private.address;
const abi = artifact.abi;

const contract = new Contract(artifact.abi);
const refund = "0x4b7b1cbbd739a2a0e95b32b64fd3d249c671bd44";


export const create = (
  paymentAddress: string,
  amount: string,
  ipfsHash: any
) => {
  const txData = contract.methods.createRequestAsPayer(
    [paymentAddress],
    [amount],
    refund,
    [0],
    [0],
    "random"
  );

  const txParams = {
    nonce: "0x00",
    //    gasPrice: "0x09184e72a000",
    //    gasLimit: "0x2710",
    to: artifact.networks.private.address,
    //    value: "0x00",
    data: txData.encodeABI()
    //    chainId: 3
  };

  const tx = new EthereumTx(txParams);
  return tx;
};
