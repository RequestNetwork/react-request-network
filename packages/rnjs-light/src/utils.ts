import requestArtifact from "requestnetworkartifacts";

const defaultNetwork = "private";
const defaultAddress = "last-RequestEthereum";

export const getArtifact = (
  network = defaultNetwork,
  address = defaultAddress
) => requestArtifact(network, address);
