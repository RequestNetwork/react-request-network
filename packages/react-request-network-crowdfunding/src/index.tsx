import Publisher from './Publisher';
import Backer from './Backer';

export { Publisher, Backer };