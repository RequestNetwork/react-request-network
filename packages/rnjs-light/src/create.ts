const Contract = require("web3-eth-contract");
import { getArtifact } from "./helpers";
import { utils } from "ethers";

// for now only work with private networks
const artifact = getArtifact();
const contract = new Contract(artifact.abi);

export interface ICreate {
  beneficiary: string;
  amount: string;
  ipfsHash: string;
  payer: string;
}

export const create = ({ beneficiary, amount, payer, ipfsHash }: ICreate) => {
  // as the beneficiary
  const txData = contract.methods.createRequestAsPayer(
    [beneficiary],
    [amount],
    payer,
    [0],
    [0],
    ipfsHash
  );

  const txParams = {
    to: artifact.networks.private.address,
    data: txData.encodeABI()
  };
  return utils.serializeTransaction(txParams);
};
