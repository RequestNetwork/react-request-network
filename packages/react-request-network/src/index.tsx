import * as React from 'react';
import { createRequest, getRequest, payRequest } from './utils';
import RequestNetworkProvider from './Provider';
import Publisher from './Publisher';
import Backer from './Backer';

export const { Provider, Consumer } = React.createContext({
  isReady: false,
  networkMismatch: false,
  currentNetwork: '',
  currentAccount: '',
  create: createRequest,
  get: getRequest,
  pay: payRequest,
});

export { Publisher, Backer };
export default RequestNetworkProvider;
