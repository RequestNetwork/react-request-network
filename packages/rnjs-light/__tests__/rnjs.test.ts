import { create } from "../src";
const add = (n1, n2) => {
  return n1 + n2;
};

describe("create a request", () => {
  it("should contain a create function", () => {
    expect(typeof create).toEqual("function");
  });
  it("should create a valid transaction", () => {
    const paymentAddress = "0x0e8d9cb9e11278ad6e2ba1ca90385c7295dc6532";
    const amount = "100000000000000000000000000000000";
    const data = {};
    expect(create(paymentAddress, amount, data)).toMatchSnapshot();
  });
  it("should get the latest nonce", () => {});
});
