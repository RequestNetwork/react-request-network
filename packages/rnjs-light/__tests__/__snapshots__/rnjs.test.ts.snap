// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create a request should create a valid transaction 1`] = `"0xf9024080808094f12b5dd4ead5f743c6baa640b0216200e89b60da80b902241e24e02900000000000000000000000000000000000000000000000000000000000000c000000000000000000000000000000000000000000000000000000000000001000000000000000000000000004b7b1cbbd739a2a0e95b32b64fd3d249c671bd440000000000000000000000000000000000000000000000000000000000000140000000000000000000000000000000000000000000000000000000000000018000000000000000000000000000000000000000000000000000000000000001c00000000000000000000000000000000000000000000000000000000000000001000000000000000000000000627306090abab3a6e1400e9345bc60c78a8bef57000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000640000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002e516d5234506a4c3971737a42417a64526976573539416d386859526a32416b446b4537465133656e7a4c69423748000000000000000000000000000000000000"`;

exports[`e2e should create a broadcastable transaction 1`] = `
Object {
  "chainId": 0,
  "data": "0x1e24e02900000000000000000000000000000000000000000000000000000000000000c000000000000000000000000000000000000000000000000000000000000001000000000000000000000000004b7b1cbbd739a2a0e95b32b64fd3d249c671bd440000000000000000000000000000000000000000000000000000000000000140000000000000000000000000000000000000000000000000000000000000018000000000000000000000000000000000000000000000000000000000000001c00000000000000000000000000000000000000000000000000000000000000001000000000000000000000000627306090abab3a6e1400e9345bc60c78a8bef57000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000d3c21bcecceda10000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002e516d5234506a4c3971737a42417a64526976573539416d386859526a32416b446b4537465133656e7a4c69423748000000000000000000000000000000000000",
  "from": "0x627306090abaB3A6e1400e9345bC60c78a8BEf57",
  "gasLimit": BigNumber {
    "_hex": "0x00",
  },
  "gasPrice": BigNumber {
    "_hex": "0x00",
  },
  "hash": "0x99842fe29a7a9ad7ecd1cb60c894ee6ecc85621bd56cb71019bf341ff417a551",
  "nonce": 33,
  "r": "0x064066c647781b5e7601cb14d77e9c077ff80c362dd1b4cb1087f3f490baf042",
  "s": "0x125e7889911b6a2fc1682384d1a89cc30b77e6abb6a12d982d70260d7a65813a",
  "to": "0xF12b5dd4EAD5F743C6BaA640B0216200e89B60Da",
  "v": 27,
  "value": BigNumber {
    "_hex": "0x00",
  },
}
`;

exports[`e2e should parse a signed transaction 1`] = `
Object {
  "chainId": 0,
  "data": "0x1e24e02900000000000000000000000000000000000000000000000000000000000000c000000000000000000000000000000000000000000000000000000000000001000000000000000000000000004b7b1cbbd739a2a0e95b32b64fd3d249c671bd440000000000000000000000000000000000000000000000000000000000000140000000000000000000000000000000000000000000000000000000000000018000000000000000000000000000000000000000000000000000000000000001c00000000000000000000000000000000000000000000000000000000000000001000000000000000000000000627306090abab3a6e1400e9345bc60c78a8bef57000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000640000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002e516d5234506a4c3971737a42417a64526976573539416d386859526a32416b446b4537465133656e7a4c69423748000000000000000000000000000000000000",
  "from": "0x627306090abaB3A6e1400e9345bC60c78a8BEf57",
  "gasLimit": BigNumber {
    "_hex": "0x00",
  },
  "gasPrice": BigNumber {
    "_hex": "0x00",
  },
  "hash": "0x8987c20cf0de63ec30e52b8dff70e86fddffa24140541eb028700ac5cbc994e8",
  "nonce": 0,
  "r": "0x7581ee61890d57a6ed257a61acbfebc4a647333001cb72d50a0ad26b63b83678",
  "s": "0x1b5df879e3f12a0da7b7c9a4f82cff660650af7850459664a94b2be722c916b1",
  "to": "0xF12b5dd4EAD5F743C6BaA640B0216200e89B60Da",
  "v": 28,
  "value": BigNumber {
    "_hex": "0x00",
  },
}
`;
